File: core/src/main/java/com/xmartlabs/base/core/module/GsonExclusionStrategy.java
package com.xmartlabs.base.core.module;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.annimon.stream.Objects;
import com.annimon.stream.Optional;
import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.xmartlabs.base.core.common.GsonExclude;

import java.util.List;

import lombok.RequiredArgsConstructor;

/** Provides a {@link ExclusionStrategy} that excludes {@link GsonExclude} fields and any other specified class. */
@RequiredArgsConstructor
public class GsonExclusionStrategy {
  private final List<Class<?>> excludedClasses;

  /**
   * Retrieves an {@link ExclusionStrategy} that excludes {@link GsonExclude} fields and the classes contained in
   * {@code excludedClasses}.
   *
   * @param strategy the type of the strategy to be retrieved
   * @return the {@link ExclusionStrategy} for the {@code strategy} provided
   */
  public ExclusionStrategy getExclusionStrategy(@Nullable GsonExclude.Strategy strategy) {
    return new ExclusionStrategy() {
      @Override
      public boolean shouldSkipField(FieldAttributes fieldAttributes) {
        return shouldSkipFieldFromSerialization(fieldAttributes)
            || (fieldAttributes.getAnnotation(GsonExclude.class) != null
            && (Objects.equals(fieldAttributes.getAnnotation(GsonExclude.class).strategy(), GsonExclude.Strategy.ALL)
            || Objects.equals(fieldAttributes.getAnnotation(GsonExclude.class).strategy(), strategy)));
      }

      @Override
      public boolean shouldSkipClass(Class<?> clazz) {
        return false;
      }
    };
  }

  /**
   * Returns whether or not the field with {@code fieldAttributes} should be serialized.
   *
   * @param fieldAttributes the attributes of the field to analise
   * @return whether or not the field should be serialized
   */
  protected boolean shouldSkipFieldFromSerialization(@NonNull FieldAttributes fieldAttributes) {
    return Optional.ofNullable(excludedClasses)
        .map(classes -> classes.contains(fieldAttributes.getDeclaredClass()))
        .orElse(false);
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/module/GsonExclusionStrategy.kt
package com.xmartlabs.bigbang.core.module

import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.xmartlabs.bigbang.core.common.GsonExclude
import com.xmartlabs.bigbang.core.common.GsonExcludeStrategy

/** Provides a [ExclusionStrategy] that excludes [GsonExclude] fields and any other specified class.  */
open class GsonExclusionStrategy(val excludedClasses: List<Class<Any>>? = null) {
  /**
   * Retrieves an [ExclusionStrategy] that excludes [GsonExclude] fields and the classes contained in
   * `excludedClasses`.
   * @param strategy the type of the strategy to be retrieved
   * *
   * @return the [ExclusionStrategy] for the `strategy` provided
   */
  open fun getExclusionStrategy(strategy: GsonExcludeStrategy?) = object : ExclusionStrategy {
    override fun shouldSkipField(fieldAttributes: FieldAttributes) =
        shouldSkipFieldFromSerialization(fieldAttributes)
            || fieldAttributes.getAnnotation(GsonExclude::class.java) != null
            && (fieldAttributes.getAnnotation(GsonExclude::class.java).strategy == GsonExcludeStrategy.ALL
            || fieldAttributes.getAnnotation(GsonExclude::class.java).strategy == strategy)
  
    override fun shouldSkipClass(clazz: Class<*>) = false
  }

  /**
   * Returns whether or not the field with `fieldAttributes` should be serialized.
   *
   * @param fieldAttributes the attributes of the field to analise
   * *
   * @return whether or not the field should be serialized
   */
  protected open fun shouldSkipFieldFromSerialization(fieldAttributes: FieldAttributes) =
      excludedClasses?.contains(fieldAttributes.declaredClass) ?: false
}
